syft/formats/common/spdxhelpers/to_syft_model.go

```diff
@@ -49,6 +51,125 @@ func ToSyftModel(doc *spdx.Document) (*sbom.SBOM, error) {
 	return s, nil
 }
 
+func isDirectory(name string) bool {
+	if name == "." || name == ".." || strings.HasSuffix(name, "/") || !strings.Contains(path.Base(name), ".") {
+		return true
+	}
+	return false
+}
+
+func removePackage(packages []*spdx.Package, remove *spdx.Package) (pkgs []*spdx.Package) {
+	for _, p := range packages {
+		if p == remove {
+			continue
+		}
+		pkgs = append(pkgs, p)
+	}
+	return
+}
+
+func removeRelationships(relationships []*spdx.Relationship, spdxID spdx.ElementID) (relations []*spdx.Relationship) {
+	for _, r := range relationships {
+		if r.RefA.ElementRefID == spdxID || r.RefB.ElementRefID == spdxID {
+			continue
+		}
+		relations = append(relations, r)
+	}
+	return
+}
+
+func findRootPackages(doc *spdx.Document) (out []*spdx.Package) {
+	for _, p := range doc.Packages {
+		for _, r := range doc.Relationships {
+			if r.RefA.ElementRefID != "DOCUMENT" ||
+				r.Relationship != spdx.RelationshipDescribes ||
+				r.RefB.ElementRefID != p.PackageSPDXIdentifier {
+				continue
+			}
+			out = append(out, p)
+		}
+	}
+	return
+}
+
+//nolint:funlen
+func extractSource(spdxIDMap map[string]any, doc *spdx.Document) source.Description {
+	src := extractSourceFromNamespace(doc.DocumentNamespace)
+
+	rootPackages := findRootPackages(doc)
+
+	if len(rootPackages) != 1 {
+		return src
+	}
+
+	p := rootPackages[0]
+
+	typeRegex := regexp.MustCompile("^DocumentRoot-([^-]+)-.*$")
+	typeName := typeRegex.ReplaceAllString(string(p.PackageSPDXIdentifier), "$1")
+
+	id := string(p.PackageSPDXIdentifier)
+	switch p.PrimaryPackagePurpose {
+	case spdxPrimaryPurposeContainer:
+		container := p.PackageName
+		v := p.PackageVersion
+		if v != "" {
+			container += ":" + v
+		}
+		if len(p.PackageChecksums) > 0 {
+			c := p.PackageChecksums[0]
+			id = fmt.Sprintf("%s:%s", fromChecksumAlgorithm(c.Algorithm), c.Value)
+		}
+		src = source.Description{
+			ID:      id,
+			Name:    p.PackageName,
+			Version: p.PackageVersion,
+			Metadata: source.StereoscopeImageSourceMetadata{
+				UserInput: container,
+				ID:        id,
+				Layers:    nil, // TODO handle formats with nested layer packages like Tern and K8s BOM tool
+			},
+		}
+	case spdxPrimaryPurposeFile:
+		version := p.PackageVersion
+		var metadata any = source.DirectorySourceMetadata{
+			Path: p.PackageName,
+			Base: "",
+		}
+		if typeName != prefixDirectory && (typeName == prefixFile || !isDirectory(p.PackageName)) {
```

Comment:
Why is this case needed if the prefix explicitly states file? ` !isDirectory(p.PackageName)`

That is, the `isDirectory` is trying to lexically get to an answer that I don't think is right in all cases but there is already an answer encoded next to `DocumentRoot-...`. I feel I'm missing something here.

Reply:
Based on https://github.com/anchore/syft/pull/1934/files#r1263105669 I think we should try to only use the lexical detection in cases where we think this wasn't generated by syft. After looking at this section for a while I think it does do exactly that, but the code really doesn't make this obvious.

Something like this would make more apparent what higher-level logic this is representing:
```golang

isFromSyft := typeName == prefixDirectory || typeName == prefixFile

if isFromSyft {
  switch typeName {
    case prefixDirectory:
      // make directory metadata...
    case prefixFile:
      // make file metadata...
  }
} else {
  // lexically detect if it's a directory or file
}

```

this also motivates breaking some of the metadata creation sections into separate functions for reuse (and makes this function smaller without needing a linter ignore).